lib/src/main/java/graphql/nadel/NadelResultMerger.kt

```diff
@@ -0,0 +1,156 @@
+package graphql.nadel
+
+import graphql.ExecutionResult
+import graphql.ExecutionResultImpl
+import graphql.GraphQLError
+import graphql.introspection.Introspection
+import graphql.nadel.engine.transform.query.NadelFieldAndService
+import graphql.nadel.engine.util.AnyMap
+import graphql.nadel.engine.util.JsonMap
+import graphql.nadel.engine.util.MutableJsonMap
+import graphql.nadel.engine.util.asJsonMap
+import graphql.nadel.engine.util.asMutableJsonMap
+import graphql.nadel.engine.util.isNonNull
+import graphql.nadel.util.ErrorUtil
+import graphql.nadel.util.NamespacedUtil.isNamespacedField
+import graphql.normalized.ExecutableNormalizedField
+import graphql.schema.GraphQLSchema
+
+@JvmInline
+private value class ResultKey(val value: String)
+
+internal class NadelResultMerger {
+    fun mergeResults(
+        fields: List<NadelFieldAndService>,
+        engineSchema: GraphQLSchema,
+        results: List<ServiceExecutionResult>,
+    ): ExecutionResult {
+        val data: MutableJsonMap = mutableMapOf()
+        val extensions: MutableJsonMap = mutableMapOf()
+        val errors: MutableList<GraphQLError> = mutableListOf()
+
+        for (result in results) {
+            putAndMergeTopLevelData(data, oneData = result.data)
+            errors.addAll(ErrorUtil.createGraphQLErrorsFromRawErrors(result.errors))
+            extensions.putAll(result.extensions)
+        }
+
+        return ExecutionResultImpl.newExecutionResult()
+            .data(fixData(fields, engineSchema, data))
+            .extensions(extensions.let {
+                @Suppress("UNCHECKED_CAST") // .extensions should take in a Map<*, *> instead of strictly Map<Any?, Any?>
+                it as Map<Any?, Any?>
+            }.takeIf {
+                it.isNotEmpty()
+            })
+            .errors(errors)
+            .build()
+    }
+
+    private fun fixData(
+        fields: List<NadelFieldAndService>,
+        engineSchema: GraphQLSchema,
+        data: MutableJsonMap,
+    ): MutableJsonMap? {
+        val requiredFieldMap = buildRequiredFieldMap(fields, engineSchema)
```

Comment:
So what this does is return a list of fields that should be in the result e.g. given a query

```graphql
query {
  projects {
    project(id: "1") {
      key
    }
    board(id: "1") {
      id
    }
  }
  issueById(id: "1") {
    key
  }
}
```

Then if `projects` is not null then we need both `project` and `board` in the result i.e. the following is a _valid_ result

```json
{
  "data": {
    "projects": {
      "project": {
        "key": "TEST"
      },
      "board": {
        "id": "1"
      }
    },
    "issueById": {
      "key": "TEST-1"
    }
  }
}
```

But, the following are not.

```json
{
  "data": {
    "projects": {
      "project": {
        "key": "TEST"
      }
    },
    "issueById": {
      "key": "TEST-1"
    }
  }
}
```

(above is missing `projects.board`)

```json
{
  "data": {
    "projects": {
      "board": {
        "id": "1"
      }
    },
    "issueById": {
      "key": "TEST-1"
    }
  }
}
```

(above is missing `projects.project`)

```json
{
  "data": {
    "projects": {
      "project": {
        "key": "TEST"
      },
      "board": {
        "id": "1"
      }
    }
  }
}
```

(above is missing `issueById`)

etc.

The `requiredFieldMap` for this query is

```json
{
  "projects": [
    "project",
    "board"
  ],
  "issueById": []
}
```